rfc/20240929-opentelemetry-improvements.md

```diff
@@ -0,0 +1,148 @@
+
+An example of this may be exporting to a jaeger instance that you have started using docker:
+
+```shell
+ docker run \       
+  --rm \
+  --name jaeger \
+  -e COLLECTOR_OTLP_ENABLED=true \
+  -p 16686:16686 \
+  -p 4317:4317 \
+  -p 4318:4318 \
+  jaegertracing/all-in-one:1.54.0
+```
+
+And then setting the environment variables to be used by OpenTofu:
+
+```shell
+export OTEL_EXPORTER=otel
+export OTEL_SERVICE_NAME=opentofu
+export OTEL_EXPORTER_OTLP_ENDPOINT=localhost:4317
+tofu init
+```
+
+### Technical Approach
+
+For each package in which we intend to produce traces from, we introduce a new package-wide variable that will hold the `Tracer` object. This will allow us to create a new `Tracer` object for each package that we want to produce traces from.
+
+For example, you can see the existing `telemetry.go` in the `internal/command` package:
+
+```go
+...
+package command
+
+import (
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/trace"
+)
+
+var tracer trace.Tracer
+
+func init() {
+	tracer = otel.Tracer("github.com/opentofu/opentofu/internal/command")
+}
+```
+
+This `tracer` object can then be used to create spans in the codebase. For example:
+```go
+	ctx, span := tracer.Start(ctx, "install providers")
+	defer span.End()
+```
+
+As you can see, the tracer.Start call will require the `context.Context` object to be passed in. This is the main reason why we need to refactor the codebase to support passing around the `context.Context` object. Without this refactoring, we will not be able to produce decent traces from the codebase.
+
+#### The fun part: Adding context.Context everywhere
+
+The main part of this RFC will be to refactor the codebase to support passing around the `context.Context` object. This will be a large amount of work, and will require a lot of testing to ensure that we have not broken anything in the codebase.
+
+In situations where a context is already passed into a method, it is usually to ensure that we're timing out a call after a certain amount of time. In these situations, I propose that we pass a second context to the method called `traceCtx` which is used for just tracing. Whilst this does introduce a little bit of complexity, it reduces the risk of the initial implementation by not requiring us to merge the `context.Context` object with the `traceCtx` object.
+
+#### How to handle different use cases?
+We may find that we want to provide 2 different levels of tracing, one for end-user consumption and one for debugging and internal use. In this case,
+we may want to have a way to turn some certain tracing on/off.
+
+I propose we have 2 different levels of tracing. This will mean that instead of exposing a `Tracer` object in the package, we instead expose something that is an extremely thin wrapper that understands if we are in "debug" tracing mode, or just "user" tracing mode.
```

Comment:
We need to pay additional attention to naming here since it could be confusing to have `LOG_LEVEL=TRACE` with `TRACING_MODE=DEBUG`.

ps I assume we don't want to map logging levels with tracing modes